internal/pkg/instrumentation/bpf/go.opentelemetry.io/otel/global/bpf/probe.bpf.c

```diff
@@ -0,0 +1,173 @@
+    char buf[MAX_SPAN_NAME_LEN];
+};
+
+struct otel_span_t {
+    BASE_SPAN_PROPERTIES
+    struct span_name_t span_name;
+    otel_attributes_t attributes;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, void*);
+	__type(value, struct otel_span_t);
+	__uint(max_entries, MAX_CONCURRENT);
+} active_spans_by_span_ptr SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, void*);
+	__type(value, struct span_name_t);
+	__uint(max_entries, MAX_CONCURRENT);
+} span_name_by_context SEC(".maps");
+
+struct
+{
+    __uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+    __uint(key_size, sizeof(u32));
+    __uint(value_size, sizeof(struct otel_span_t));
+    __uint(max_entries, 2);
+} otel_span_storage_map SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+} events SEC(".maps");
+
+// Injected in init
+volatile const u64 span_name_pos;
+volatile const u64 span_attributes_pos;
+
+// This instrumentation attaches uprobe to the following function:
+// func (t *tracer) Start(ctx context.Context, name string, opts ...trace.SpanStartOption) (context.Context, trace.Span)
+SEC("uprobe/Start")
+int uprobe_Start(struct pt_regs *ctx) {
+    struct span_name_t span_name = {0};
+
+    // Getting span name
+    void *span_name_ptr = get_argument(ctx, 4);
+    u64 span_name_len = (u64)get_argument(ctx, 5);
+    u64 span_name_size = MAX_SPAN_NAME_LEN < span_name_len ? MAX_SPAN_NAME_LEN : span_name_len;
+    bpf_probe_read(span_name.buf, span_name_size, span_name_ptr);
+
+    // Save the span name in map to be read once the Start function returns
+    void *context_ptr_val = get_Go_context(ctx, 3, 0, true);
+    void *key = get_consistent_key(ctx, context_ptr_val);
+    bpf_map_update_elem(&span_name_by_context, &key, &span_name, 0);
+    return 0;
+}
+
+// This instrumentation attaches uprobe to the following function:
+// func (t *tracer) Start(ctx context.Context, name string, opts ...trace.SpanStartOption) (context.Context, trace.Span)
+SEC("uprobe/Start")
+int uprobe_Start_Returns(struct pt_regs *ctx) {
+    // Get the span name passed to the Start function
+    void *context_ptr_val = get_Go_context(ctx, 3, 0, true);
+    void *key = get_consistent_key(ctx, context_ptr_val);
+    struct span_name_t *span_name = bpf_map_lookup_elem(&span_name_by_context, &key); 
+    if (span_name == NULL) {
+        return 0;
+    }
+    bpf_map_delete_elem(&span_name_by_context, &key);
+
+    u32 zero_span_key = 0;
+    struct otel_span_t *zero_span = bpf_map_lookup_elem(&otel_span_storage_map, &zero_span_key);
+    if (zero_span == NULL) {
+        return 0;
+    }
+
+    u32 otel_span_key = 1;
+    // Zero the span we are about to build, eBPF doesn't support memset of large structs (more than 1024 bytes)
+    bpf_map_update_elem(&otel_span_storage_map, &otel_span_key, zero_span, 0);
+    // Get a pointer to the zeroed span
+    struct otel_span_t *otel_span = bpf_map_lookup_elem(&otel_span_storage_map, &otel_span_key);
```

Comment:
Was the [C++ OTel implementation](https://github.com/open-telemetry/opentelemetry-cpp/tree/main/sdk/src/trace) considered for use here instead?

Reply:
I don't think it is possible since eBPF is run inside the kernel, and the allowed functions to call is quite limited https://ebpf.io/what-is-ebpf/#helper-calls